content/advent-2016/enigma-emulator-in-go.md

```diff
@@ -0,0 +1,486 @@
+
+Let's create a simple type for the rotors:
+
+```
+type Rotor struct {
+	ID          string
+	StraightSeq [26]int
+	ReverseSeq  [26]int
+	Turnover    []int
+
+	Offset int
+	Ring   int
+}
+
+type Rotors []*Rotor
+```
+
+Note that letter sequences are `[26]int`: inside the library, we are going to be working with `int` values for letters (`0` for `'A'`, `25` for `'Z'`), rather than bytes or runes. We do not care about the characters themselves when performing the encoding — their position and offset matter much more — so we are saving ourselves from a bunch of extra `rune`/`int` conversions during the encoding process.
+
+We still have to perform the conversion in constructors, so here are the two helper functions:
+
+```
+func CharToIndex(char byte) int {
+	return int(char - 'A')
+}
+
+func IndexToChar(index int) byte {
+	return byte('A' + index)
+}
+```
+
+Curiously, the reason something like this works is that `rune` is an alias for `int32` in Go; there is [a useful post in The Go Blog](https://blog.golang.org/strings) on characters, bytes, and strings that I would recommend.
+
+Let's define a constructor and create the eight preset rotors that were in use at the time. A helper method for the rotor list to get rotors by ID is included (although making `Rotors` a map instead of a list is also an option):
+
+```
+func NewRotor(mapping string, id string, turnovers string) *Rotor {
+	r := &Rotor{ID: id, Offset: 0, Ring: 0}
+	r.Turnover = make([]int, len(turnovers))
+	for i := range turnovers {
+		r.Turnover[i] = CharToIndex(turnovers[i])
+	}
+	for i, letter := range mapping {
+		index := CharToIndex(byte(letter))
+		r.StraightSeq[i] = index
+		r.ReverseSeq[index] = i
+	}
+	return r
+}
+
+func (rs *Rotors) GetByID(id string) *Rotor {
+	for _, rotor := range *rs {
+		if rotor.ID == id {
+			return rotor
+		}
+	}
+	return nil
+}
+
+var HistoricRotors = Rotors{
+	NewRotor("EKMFLGDQVZNTOWYHXUSPAIBRCJ", "I", "Q"),
+	NewRotor("AJDKSIRUXBLHWTMCQGZNPYFVOE", "II", "E"),
+	NewRotor("BDFHJLCPRTXVZNYEIWGAKMUSQO", "III", "V"),
+	NewRotor("ESOVPZJAYQUIRHXLNFTGKDCMWB", "IV", "J"),
+	NewRotor("VZBRGITYUPSDNHLXAWMJQOFECK", "V", "Z"),
+	NewRotor("JPGVOUMFYQBENHZRDKASXLICTW", "VI", "ZM"),
+	NewRotor("NZJHGRCXMYSWBOUFAIVLPEKQDT", "VII", "ZM"),
+	NewRotor("FKQHTLXOCBJSPDZRAMEWNIUYGV", "VIII", "ZM")
+}
+```
+
+Finally, the method performing the substitution. We need to account for the current position and the ring settings — this is one of the examples where thinking of letters as their indexes helps us avoid unnecessary conversions:
+
+```
+func (r *Rotor) Step(letter *int, invert bool) {
```

Comment:
It was another pprof-induced change, but I'll try to make it return the letter again and see if it doesn't slow down things too much. It'd certainly be less confusing that way.